x-pack/plugins/fleet/server/services/package_policy.ts

```diff
@@ -595,48 +573,26 @@ class PackagePolicyService {
     soClient: SavedObjectsClientContract,
     esClient: ElasticsearchClient,
     ids: string[],
-    options?: { user?: AuthenticatedUser }
+    options?: { user?: AuthenticatedUser },
+    packagePolicy?: PackagePolicy,
+    pkgVersion?: string
   ): Promise<UpgradePackagePolicyResponse> {
     const result: UpgradePackagePolicyResponse = [];
 
     for (const id of ids) {
       try {
-        const { packagePolicy, packageInfo } = await this.getUpgradePackagePolicyInfo(soClient, id);
-
-        const updatePackagePolicy = updatePackageInputs(
-          {
-            ...omit(packagePolicy, 'id'),
-            inputs: packagePolicy.inputs,
-            package: {
-              ...packagePolicy.package,
-              version: packageInfo.version,
-            },
-          },
-          packageInfo,
-          packageToPackagePolicyInputs(packageInfo) as InputsOverride[]
-        );
-        const registryPkgInfo = await Registry.fetchInfo(packageInfo.name, packageInfo.version);
-        updatePackagePolicy.inputs = await this._compilePackagePolicyInputs(
-          registryPkgInfo,
-          packageInfo,
-          updatePackagePolicy.vars || {},
-          updatePackagePolicy.inputs as PackagePolicyInput[]
-        );
-        updatePackagePolicy.elasticsearch = registryPkgInfo.elasticsearch;
+        let packageInfo: PackageInfo | undefined;
+        if (!packagePolicy) {
+          ({ packagePolicy, packageInfo } = await this.getUpgradePackagePolicyInfo(soClient, id));
+        } else if (!packageInfo) {
+          packageInfo = await getPackageInfo({
```

Comment:
one thing I noticed that `getPackageInfo` is called in a for loop for each package policy id, this could be optimized to query once for each package only

Reply:
`getPackageInfo` should resolve from Fleet's in-memory cache on subsequent runs, so we might be duplicating memoization efforts here. But, it's worth looking into for sure.

Reply:
I measured this earlier locally, and it seemed that `getPackageInfo` takes about 100ms for the first time, and immediate subsequent calls take about 50ms. This still seems much if someone has 10-20 integration policies. 
Can you point me to the code where this in-memory caching is done?

Reply:
`getPackageFromSource` will resolve the package from cache if possible: 

https://github.com/elastic/kibana/blob/c94d5fdd9e61818a8c4fecdeed2ff3852a5112c3/x-pack/plugins/fleet/server/services/epm/packages/get.ts#L160-L165

https://github.com/elastic/kibana/blob/c94d5fdd9e61818a8c4fecdeed2ff3852a5112c3/x-pack/plugins/fleet/server/services/epm/packages/get.ts#L237-L247

But, we don't cache the installation object or latest package record in `getPackageInfo`, so that's likely where the added time is coming from

https://github.com/elastic/kibana/blob/c94d5fdd9e61818a8c4fecdeed2ff3852a5112c3/x-pack/plugins/fleet/server/services/epm/packages/get.ts#L147-L150

Reply:
Yes, it seems like `Registry.fetchFindLatestPackageOrUndefined(pkgName)` is the culprit, we are calling this each time even when the `packageInfo` object is cached, to check whether there is a new version of the package in the registry. 
I think this is an overkill, perhaps the cache could be made smarter to avoid calling the registry for some time e.g. 1 day, since I think it's a rare event to find new versions published.

Reply:
Yeah I agree there. I think caching the actual registry responses is a good idea. We do have this issue to track potentially honoring various `cache-control` headers on responses from EPR, but I don't know that that applies to server-side requests: https://github.com/elastic/kibana/issues/125794

I do think determining a caching strategy for our "latest package" requests is out of scope for this PR, though, so we don't need to come up with a solution in order to land this.

Reply:
Good point, cache-control definitely helps with caching Fleet API calls.
As for server side, I think that depends on node fetch capabilities, which are different than in browser.